### ISSUE TITLE
Add Exception Handler

### ISSUE THREAD

**Issue by ckdarby** [EloGank/lol-replay-downloader-cli#2]
How do you feel about this? We have a need to override the game found exception to not throw an error. Was hoping to fork & add functionality similar to the success & failure handler.

**Comment by Divi** [comment]
I don't remember if the Symfony application throws an exception, but did you try to handle the exception in the `console` file, with a try/catch on the `$application->run()` ?

**Comment by ckdarby** [comment]
Sorry, my initial message was written in a rather rush. While the try/catch option would work, I would rather the option of not needing to modify the source code of the project altogether. It causes maintainability issues when pulling upstream down in the future.

The ideal situation was similar to the handler set up with [config.yml](https://github.com/EloGank/lol-replay-downloader-cli/blob/master/config/config.yml.dist)

It would along the lines of:

```YML

        exceptions:
            handler:
                # Override exception, must be a class namespace
                GameEndedException: ~
                # Override exception, must be a class namespace
                GameNotFoundException: ~
                # Override exception, must be a class namespace
                GameNotStartedException: ~
                # Override exception, must be a class namespace
                ReplayFolderAlreadyExistsException: \Ckdarby\Elogank\Exceptions\ReplayFolderAlreadyExistsException

```

**Comment by Divi** [comment]
Can you explain me why you need to override the exception and why the failure handler cannot handle this case ?  
Because this solution seems to be a little bit tricky and I would improve the failure handler instead of implementing a class injector.

Thanks @ckdarby :)

**Comment by Divi** [comment]
Any news @ckdarby ?

**Comment by ckdarby** [comment]
@Divi The replay downloader is solid. It handles the case where it is already downloading a replay and won't attempt to download the same replay. The problem is it makes our logs totally useless because of the spam caused from ReplayFolderAlreadyExistsException.

We'd like a way to suppress this. There's no reason to care about the exception. I don't want to modify the existing code base because this is the same logic around the success & failure handlers; There's a reason to have the class injectors instead.

TL;DR CLI tool floods our syslogs

**Comment by Divi** [comment]
@ckdarby, I understand but the main purpose of this CLI is not to validate the existence of a replay. Maybe you should implement a validation before running the command by passing the ID in Redis, for example.

On the other hand, if you can't create a middleware, I suggest you to check the instance of the exception in your failure handler, and run a `die()` if it's an instance of ReplayFolderAlreadyExistsException. The exception won't be thrown in this case.

How do you feel about this suggestion ?

**Comment by ckdarby** [comment]
I agree with using the onFailure handler if it worked as expected; Writing a die() statement in the failure handler is the wrong solution in my opinion because it doesn't allow the application to gracefully end but instead I am just terminating it.

https://github.com/EloGank/lol-replay-downloader-cli/blob/master/src/EloGank/Replay/Command/ReplayDownloadCommand.php#L106-L111

I'm not all that familiar with the Symfony console component but I am confused to why you're throwing the exception up. There isn't anything else that should be wrapping the CLI application. If it didn't throw the exceptions up then I'd be able to use the onFailure handler and not need to terminate with die().

**Comment by Divi** [comment]
You're right. I let the exception mainly for development purpose and let me retrieve the full stack.
**Here another suggestion :** let the choice to the failure handler to throw or not the exception by returning `true` or `false` ?

If your handler returns `true`, the command will throw the exception, allowing a developper with `dev` environment to have the full stack in the logs. In the other hand, for a `production` env like your, you can just return `false` to bypass the error, a quiet mode.

**Comment by ckdarby** [comment]
Bump out to config file instead?

```PHP

if (Config::get('replay.dev_mode')) {
 throw $e;
}
```

**Comment by Divi** [comment]
No, some exceptions are useful, like an invalid game for example.
I must let the choice to the develop to select which exception he wants to handle.

You just have to create your FailureHandler like this :

``` php
public function onFailure($region, $gameId, $encryptionKey, $replayFolderPath, \Exception $exception)
{
    if ($exception instance of ReplayFolderAlreadyExistsException) {
        $this->myLogger->error('Wow, such bad');

        return false;
    }
}
```

**Comment by ckdarby** [comment]
I totally agree but that is already solved in the FailureHandler. What I am discussing is outside of the FailureHandler.

https://github.com/EloGank/lol-replay-downloader-cli/blob/master/src/EloGank/Replay/Command/ReplayDownloadCommand.php#L106-L111

Change the throw $e to

```PHP

if (Config::get('replay.dev_mode')) {
 throw $e;
}
```

The developer is still able to select which exceptions they want to handle within the FailureHandler but throwing the exception up outside of the CLI application only happens if the config is set for it. 

> You're right. I let the exception mainly for development purpose and let me retrieve the full stack.

There's allows a toggle of on/off through the config for them to retrieve the full stack.

**Comment by Divi** [comment]
:+1:

**Comment by ckdarby** [comment]
:+1:

Are you going to add this or did you want me to PR it?

**Comment by Divi** [comment]
I'm going to add this, let me a couple of minutes :)

**Comment by Divi** [comment]
@ckdarby In your syslog, do you have the full stack ?

**Comment by ckdarby** [comment]
@Divi I don't even remember. I just remember us entirely turning off logging for the CLI stuff because of this.

**Comment by Divi** [comment]
Ok, see https://github.com/EloGank/lol-replay-downloader-cli/commit/5316f06c66bb652499b6773a639556b45a4f1d74 (on the branch `feature-exception`). If it's ok, I'll merge it tomorrow.

If the `replay.command.exception.throw` is set to `false`, a simple message will be shown (like "The replay #123456789 already exists").

If your syslog is still flooded because of this simple message, I suggest you to implement a middleware to validate if a replay already exists.

**Comment by Divi** [comment]
@ckdarby any news ?

**Comment by ckdarby** [comment]
Swapped to origin/feature-exception branch. 


```BASH
php /home/gamerbet/lol-replay-downloader-cli/src/EloGank/Replay/Command/../../../../console elogank:replay:download NA1 1999408138 4LstWSAin07QIJ6c4A2PuFAZjulJyF2

PHP Fatal error:  Cannot inherit previously-inherited or override constant VERBOSITY_QUIET from interface Symfony\Component\Console\Output\OutputInterface in /home/gamerbet/lol-replay-downloader-cli/src/EloGank/Replay/Command/Output/ConsoleOutput.php on line 22
oleOutput.php on line 22
```

**Comment by Divi** [comment]
Sorry @ckdarby, I commited by mistake these constants. Pull the branch, run a `composer update elogank*` and it should work. Tested on my VM.

**Comment by ckdarby** [comment]
@Divi Looks good. :+1:

**Comment by ckdarby** [comment]
I'd like to continue a conversation with you on where you plan to go with all of this. What is the best way to reach you?

**Comment by Divi** [comment]
@ckdarby Great, thanks for the feedback, I'll make the merge tonight.
Contact me by email, it's available on this project files header or on my Github profile :+1:

**Comment by Divi** [comment]
Sorry for the late. The branch has been merged into `master`. If I have the time, I'll make the composer release tonight, otherwise it will be tomorrow.

Thanks again @ckdarby :+1:
